R/remote.R

```diff
@@ -242,83 +201,74 @@ get_default_remote <- function(config = NULL, locate = TRUE) {
-    assert_scalar(remote)
-    if (remote %in% names(config$api_server)) {
-      check_remote_api_server(config, config$api_server[[remote]])
-    } else if (file.exists(remote)) {
-      orderly_remote_path(remote)
-    } else {
-      stop(sprintf("Unknown remote '%s'", remote),
-           call. = FALSE)
-    }
-  } else {
+    return(get_default_remote(config))
+  }
+  if (implements_remote(remote)) {
+    return(remote)
+  }
+  if (!is.character(remote)) {
     stop("Unknown remote type ",
          paste(squote(class(remote)), collapse = " / "),
          call. = FALSE)
   }
+
+  assert_scalar(remote)
+  if (remote %in% names(config$remote)) {
+    load_remote(remote, config)
+  } else if (file.exists(remote)) {
+    orderly_remote_path(remote)
+  } else {
+    stop(sprintf("Unknown remote '%s'", remote), call. = FALSE)
+  }
 }
 
 
-remote_report_names <- function(config = NULL, locate = TRUE, remote = NULL) {
-  config <- orderly_config_get(config, locate)
-  remote <- get_remote(remote, config)
-
-  check_remote_type(remote)
-  if (inherits(remote, "montagu_server")) {
-    remote_report_names_api(remote)
-  } else if (inherits(remote, "orderly_remote_path")) {
-    remote_report_names_path(remote)
+load_remote <- function(name, config) {
+  remote <- config$remote[[name]]
+  hash <- hash_object(remote)
+  if (is.null(cache$remotes[[hash]])) {
+    driver <- getExportedValue(remote$driver[[1L]], remote$driver[[2L]])
+    args <- resolve_secrets(remote$args, config)
+    cache$remotes[[hash]] <- do.call(driver, args)
   }
+  cache$remotes[[hash]]
+}
+
+
+## Most of these functions can really shrink now?
+remote_report_names <- function(config = NULL, locate = TRUE, remote = NULL) {
+  remote <- get_remote(remote, orderly_config_get(config, locate))
+  remote$list()
 }
 
 
 remote_report_versions <- function(name, config = NULL, locate = TRUE,
                                    remote = NULL) {
-  config <- orderly_config_get(config, locate)
-  remote <- get_remote(remote, config)
-
-  check_remote_type(remote)
-  if (inherits(remote, "montagu_server")) {
-    remote_report_versions_api(name, remote)
-  } else if (inherits(remote, "orderly_remote_path")) {
-    remote_report_versions_path(name, remote)
-  }
+  remote <- get_remote(remote, orderly_config_get(config, locate))
+  remote$list_versions(name)
 }
 
 
-check_remote_type <- function(remote) {
-  if (inherits(remote, "montagu_server") ||
-      inherits(remote, "orderly_remote_path")) {
-    return(NULL)
-  }
-  stop("Unknown remote type ",
-       paste(squote(class(remote)), collapse = " / "))
+remote_name <- function(remote) {
+  remote$name
 }
 
 
-remote_name <- function(remote) {
-  check_remote_type(remote)
-  if (inherits(remote, "montagu_server")) {
-    remote$name
-  } else if (inherits(remote, "orderly_remote_path")) {
-    as.character(remote)
-  }
+## Test to see if something (might) implement our remote interface
+implements_remote <- function(x) {
```

Comment:
When are we expecting to have to deal with objects that aren't the defined R6 classes?
Isn't `x$class == orderly_remote_path` safer?

Reply:
See the object created in https://github.com/vimc/montagu-r/pull/13 - that will confirm to the specification here but not have that class name.

In an ideal world R6 would have the concept of an interface (see, e.g., https://github.com/r-lib/R6/issues/163, https://github.com/r-lib/R6/issues/82). This function tries to bodge around that to see if we have all the methods that are expected. Ideally we'd check more about the function signature too but it's a bit dull to do.

But the core idea is that `orderly` defines what is needed to be supplied and the other packages can implement things that `orderly` does not know about without having to modify `orderly` itself - I think in OOP circles this is the [Open-Closed Principle](https://en.wikipedia.org/wiki/Open%E2%80%93closed_principle)